design/baremetal-operator/maintenance-mode.md

```diff
@@ -0,0 +1,93 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Maintenance Mode flag
+
+## Status
+
+implementable
+
+## Summary
+
+Currently there is no way to prevent management of BaremetalHost resources
+after provisioning is completed.  The underlying Ironic API contains a
+maintenance-mode flag which could be used to disable management of
+BMH resources.
+
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+This may also be useful for temporary maintenance operations
+where you may wish for power management operations and polling
+to be stopped for specific hosts (this is the intended use if the
+underlying [Ironic API](https://docs.openstack.org/api-ref/baremetal/?expanded=set-maintenance-flag-detail#set-maintenance-flag)
+
+### Goals
+
+Add a maintenance flag to the BMH API which uses the underlying Ironic
+interface to disable post-deploy management of a host.
+
+### Non-Goals
+
+
+
+
+## Proposal
+
+
+#### Baremetal Operator
+
+Introduce a new field `maintenance` in the spec of the
+BaremetalHost object. This will be a boolean flag, which
+when set to `true` will put the underlying Ironic host
+into maintenance mode.
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-0
+spec:
+  online: true
+  maintenance: true
+ ...
+```
+
+TODO - we don't expose any reason here, is it enough to have
+a standard reason string e.g "disabled by BaremetalHost maintenance field"?
```

Comment:
I assume the caller knows why the host is in maintenance mode. We could ask for a string reason, but I'm not sure what we would do with it. A UI or something could always use an annotation to store a message. Unless maybe we would show the reason as part of the default `kubectl` output for hosts?

Reply:
The underlying Ironic API does accept a reason, but I guess the user isn't exposed to that - I was thinking maybe generating an event with the reason could be useful, but it could also just be a generic event saying the host got put into maintenance mode (which simplifies the API as it's just a boolean flag)

Reply:
Ah, yeah, I wasn't really concerned with the value in Ironic so much as what we ask for or show on the host resource. Let's leave the reason out for now, under the theory that it's easier to add the field later if we decide we need it.